File: pset6_2018/vigenere/vigenere.py
import sys
import string
from cs50 import get_string


def main():
    check_args()
    key = [ord(x.upper()) - 65 for x in sys.argv[1]]
    plaintext = prompt()
    ciphertext = iterate(plaintext, key)
    print(f'ciphertext: {ciphertext}')


def check_args():
    if len(sys.argv) != 2:
        print('error: expected two command line arguments')
        exit(1)
    if not sys.argv[1].isalpha():
        print('key can only contain letters')
        exit(1)


def prompt():
    plaintext = get_string('plaintext: ')
    return plaintext


def iterate(plaintext, key):
    pos = 0
    ciphertext = ''
    for char in plaintext:
        if char.isalpha():
            ciphertext += shift(char, key[pos % len(key)])
            pos += 1
        else: 
            ciphertext += char
    return ciphertext


# source: https://stackoverflow.com/a/8895517/9352652


def shift(plaintext, key):
    alpha_low = string.ascii_lowercase
    shift_low = alpha_low[key:] + alpha_low[:key]
    table_low = str.maketrans(alpha_low, shift_low)
    alpha_up = string.ascii_uppercase
    shift_up = alpha_up[key:] + alpha_up[:key]
    table_up = str.maketrans(alpha_up, shift_up)
    cipher_low = plaintext.translate(table_low)
    return cipher_low.translate(table_up)
    

if __name__ == '__main__':
    main()

File: pset2_2017/vigenere/vigenere.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

void encrypt(char lower, char upper, int cipher, char letter);

int main(int argc, string argv[])
{
    if (argc != 2)
    {
        printf("Error: expected two command line inputs\n");
        return 1;
    }

    int len_key = strlen(argv[1]);
    for (int spot = 0; spot < len_key; spot++)
    {
        if (!(isalpha(argv[1][spot])))
        {
            printf("Key can only contain letters\n");
            return 1;
        }
    }

    // get key from user input
    string buffer = argv[1];
    int key[len_key];
    for (int i = 0; buffer[i]; i++)
    {
        buffer[i] = toupper(buffer[i]);
        key[i] = buffer[i] - 65;
    }

    printf("plaintext: ");
    string plaintext = get_string();
    int len_text = strlen(plaintext);
    int pos_key = 0;

    printf("ciphertext: ");
    for (int pos = 0; pos < len_text; pos++)
    {
        if (isalpha(plaintext[pos]))
        {
            encrypt('a', 'z', key[pos_key % len_key], plaintext[pos]);
            encrypt('A', 'Z', key[pos_key % len_key], plaintext[pos]);
            pos_key++;
        }

        else
        {
            printf("%c", plaintext[pos]);
        }
    }

    printf("\n");
    return 0;
}

// use key to encrypt user plaintext within bounds
void encrypt(char lower, char upper, int cipher, char letter)
{
    if (letter >= lower && letter <= upper)
    {
        if (letter + cipher <= upper)
        {
            printf("%c", letter + cipher);
        }

        else
        {
            int remain = (letter + cipher) - (upper);
            printf("%c", lower + (remain - 1));
        }
    }
}